File: 2024/01.ts
import { readFileSync } from 'fs';
import { resolve } from 'path';

// Type definitions
type LocationPair = [number[], number[]];

// Read input file
function readInput(): string[] {
    const filePath = resolve(__dirname, '01.input');
    return readFileSync(filePath, 'utf-8')
        .trim()
        .split('\n');
}

// Parse raw locations into two arrays
function parseLocations(lines: string[]): LocationPair {
    const locationsA: number[] = [];
    const locationsB: number[] = [];
    
    lines.forEach(line => {
        const [a, b] = line.trim().split(/\s+/).map(Number);
        locationsA.push(a);
        locationsB.push(b);
    });
    
    return [locationsA, locationsB];
}

// Calculate sum of absolute differences
function calculateDiffs(locationsA: number[], locationsB: number[]): number {
    const sortedA = [...locationsA].sort((a, b) => a - b);
    const sortedB = [...locationsB].sort((a, b) => a - b);
    
    return sortedA.reduce((sum, _, index) => 
        sum + Math.abs(sortedA[index] - sortedB[index]), 0);
}

// Calculate frequency map for locations
function calculateFrequencies(locations: number[]): Map<number, number> {
    return locations.reduce((freq, location) => {
        freq.set(location, (freq.get(location) || 0) + 1);
        return freq;
    }, new Map<number, number>());
}

// Calculate similarity score
function calculateSimilarityScore(
    locationsA: number[], 
    frequenciesB: Map<number, number>
): number {
    return locationsA.reduce((score, location) => {
        const frequency = frequenciesB.get(location) || 0;
        return score + (frequency * location);
    }, 0);
}

function main(): void {
    try {
        // Read and parse input
        const lines = readInput();
        const [locationsA, locationsB] = parseLocations(lines);
        
        // Sort locations
        const sortedA = [...locationsA].sort((a, b) => a - b);
        const sortedB = [...locationsB].sort((a, b) => a - b);
        
        // Part 1: Calculate sum of differences
        const sumDiffs = calculateDiffs(sortedA, sortedB);
        console.log(`Sum of differences: ${sumDiffs}`);
        
        // Part 2: Calculate similarity score
        const frequenciesB = calculateFrequencies(sortedB);
        const similarityScore = calculateSimilarityScore(sortedA, frequenciesB);
        console.log(`Similarity score: ${similarityScore}`);
        
    } catch (error) {
        console.error('Error processing file:', error);
        process.exit(1);
    }
}

// Run the program
main(); 

File: 2024/01.py
from typing import List, Dict, Tuple
from pathlib import Path

# Read input file
def read_input() -> List[str]:
    return Path('01.input').read_text().splitlines()

# Parse raw locations into two lists
def parse_locations(lines: List[str]) -> Tuple[List[int], List[int]]:
    locations_a: List[int] = []
    locations_b: List[int] = []
    
    for line in lines:
        # Split line and convert to integers
        a, b = map(int, line.split())
        locations_a.append(a)
        locations_b.append(b)
    
    return locations_a, locations_b

# Calculate sum of absolute differences
def calculate_diffs(locations_a: List[int], locations_b: List[int]) -> int:
    locations_a_sorted = sorted(locations_a)
    locations_b_sorted = sorted(locations_b)
    
    diffs = [abs(a - b) for a, b in zip(locations_a_sorted, locations_b_sorted)]
    return sum(diffs)

# Calculate frequency map for locations_b
def calculate_frequencies(locations: List[int]) -> Dict[int, int]:
    frequencies: Dict[int, int] = {}
    for location in locations:
        frequencies[location] = frequencies.get(location, 0) + 1
    return frequencies

# Calculate similarity score
def calculate_similarity_score(locations_a: List[int], frequencies_b: Dict[int, int]) -> int:
    score = 0
    for location in locations_a:
        if location in frequencies_b:
            score += frequencies_b[location] * location
    return score

def main():
    # Read and parse input
    lines = read_input()
    locations_a, locations_b = parse_locations(lines)
    
    # Sort locations
    locations_a_sorted = sorted(locations_a)
    locations_b_sorted = sorted(locations_b)
    
    # Part 1: Calculate sum of differences
    sum_diffs = calculate_diffs(locations_a_sorted, locations_b_sorted)
    print(f"Sum of differences: {sum_diffs}")
    
    # Part 2: Calculate similarity score
    frequencies_b = calculate_frequencies(locations_b_sorted)
    similarity_score = calculate_similarity_score(locations_a_sorted, frequencies_b)
    print(f"Similarity score: {similarity_score}")

if __name__ == "__main__":
    main() 